common/core/gpudevice.cpp

```diff
@@ -652,39 +647,44 @@ void GpuDevice::InitializeHotPlugEvents(bool take_lock) {
 }
 
 void GpuDevice::HandleRoutine() {
-  thread_sync_lock_.lock();
-  HandleHWCSettings();
-
-  // Iniitialize resources to monitor external events.
-  // These can be two types:
-  // 1) We are showing splash screen and another App
-  //    needs to take the control. In this case splash
-  //    is true.
-  // 2) Another app is having control of display and we
-  //    we need to take control.
-  // TODO: Add splash screen support.
-  lock_fd_ = open("/vendor/hwc.lock", O_RDONLY);
-  if (lock_fd_ == -1) {
-    thread_sync_lock_.unlock();
-    return;
-  }
-
-  thread_sync_lock_.unlock();
-  display_manager_->IgnoreUpdates();
+  // do nothing.
+}
 
-  if (flock(lock_fd_, LOCK_EX) != 0)
-    ETRACE("Failed to wait on hwc lock.");
+void GpuDevice::HandleWait() {
 
-  close(lock_fd_);
-  lock_fd_ = -1;
+  ETRACE("Called in GpuDevice::HandleWait.");
 
-  display_manager_->ForceRefresh();
-}
+  if (use_lock_file && -1 == lock_fd_) {
+    lock_fd_ = open("/vendor/hwc.lock", O_RDONLY);
 
-void GpuDevice::HandleWait() {
-  if (lock_fd_ == -1) {
-    HWCThread::HandleWait();
+    if (-1 == lock_fd_) {
+      ETRACE("/vendor/hwc.lock doesn't exist, skip the DRM locking");
+      use_lock_file = false;
+    }
+    else {
+      if (flock(lock_fd_, LOCK_EX|LOCK_NB) != 0) {
```

Comment:
Hmm not sure I understand the difference with this new logic and previous one. Why do we see rendering before but not now ?

Reply:
The thoughts of changes are:
1) Utilize the asynchronize HWC thread for hwc.lock grabbing and maintaining HWC rendering disable -> enabled status transition in that thread. The relevant code pieces have been removed from GpuDevice::Initialize method as implementing hwc.lock exclusively lock waiting there will block whole HWC component initialization; 
2) Once HWC thread enabled (didn't call HWCThread::Exit method in GpuDevice::Initialize method), HandleWait will be the first method to be called and then HandleRoutine. It's one time logic on rendering disable -> rendering enable, so the implementation moved to HandleWait method;
3) The first time flock call with nonblocking flag is to make the HWC have a chance to know /vendor/hwc.lock locking status. If it's hold by another process, then HWC know it has to skip the rendering content into DRM. And then issue one additional flock with blocking wait flag until another process release the exclusive lock on hwc.lock.
4) Once HWC got the exclusive lock on hwc.lock, it set HWC status to enable content rendering into DRM.

Reply:
hmm k thanks